server-database/helpers_test.go

```diff
@@ -0,0 +1,80 @@
+package main
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestValidateIndent(t *testing.T) {
+	t.Run("Valid indent", func(t *testing.T) {
+		i, err := ValidateIndent("2")
+		require.NoError(t, err)
+		require.Equal(t, 2, i)
+	})
+
+	t.Run("Negative indent", func(t *testing.T) {
+		_, err := ValidateIndent("-1")
+		require.Error(t, err)
+		require.Equal(t, err, errors.New("Indent cannot be negative: -1"))
+	})
+
+	t.Run("Invalid indent", func(t *testing.T) {
+		_, err := ValidateIndent("a")
+		require.Error(t, err)
+		require.Equal(t, err, errors.New("Unable to parse indent: a"))
+	})
+}
+
+func TestValidateAltText(t *testing.T) {
+	t.Run("Valid alt text", func(t *testing.T) {
+		err := ValidateAltText("This is a title", "This is an alt text")
+		require.NoError(t, err)
+	})
+
+	t.Run("Invalid alt text", func(t *testing.T) {
+		err := ValidateAltText("This is a title", "Alt about Something else entirely")
+		require.Error(t, err)
+		require.Equal(t, errors.New("Alt text doesn't seem to be relevant to the title"), err)
+	})
+
+	t.Run("short alt text", func(t *testing.T) {
+		err := ValidateAltText("This is a title", "hello")
+		require.Error(t, err)
+		require.Equal(t, err, errors.New("Alt text must contain at least as many words as the title"))
+	})
+}
+
+func TestValidateImage(t *testing.T) {
```

Comment:
What are the pros and cons of using images from the internet for this kind of test? What could you do instead?

Reply:
the main problem is that the image is not going to be there forever, and we don't have control over availability. There are also other factors that could contribute to failing test such as lack of internet connection or firewall blocking the connection. 

Pros: We don't need to write additional code to spin up a server and serve the image. We are also testing what the app will actually be doing - making an http request over the internet and not locally.

Reply:
One thing I'd definitely recommend is splitting off the HTTP parts of the analysis from the operating-on-bytes part of the analysis.

Things like identifying the format of the image, and the dimensions, you can trivially test with just checked in image files without needing HTTP at all.

Then the HTTP parts - I'd recommend trying to implement the smallest, simplest version of this test you could which _doesn't_ hit the public internet, and comparing the two solutions. I agree that the code to spin up a server is additional, but I'd recommend looking at both and seeing _how much_ additional, and what you think you lose in terms of test fidelity by testing locally.

Reply:
Thanks, that makes perfect sense and I have now added a few separated tests.